File: src/main/java/tech/sirwellington/alchemy/http/HttpAssertions.java
/*
 * Copyright © 2026. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.http;

import com.google.gson.JsonElement;
import tech.sirwellington.alchemy.arguments.AlchemyAssertion;
import tech.sirwellington.alchemy.arguments.FailedAssertionException;

import static tech.sirwellington.alchemy.arguments.Arguments.checkThat;
import static tech.sirwellington.alchemy.arguments.assertions.BasicAssertions.notNull;
import static tech.sirwellington.alchemy.arguments.assertions.NumberAssertions.greaterThanOrEqualTo;
import static tech.sirwellington.alchemy.arguments.assertions.NumberAssertions.lessThanOrEqualTo;
import static tech.sirwellington.alchemy.arguments.assertions.StringAssertions.nonEmptyString;
import static tech.sirwellington.alchemy.arguments.assertions.StringAssertions.stringBeginningWith;

/**
 * @author SirWellington
 */
final class HttpAssertions
{

    private HttpAssertions()
    {
        throw new AssertionError("non-instantiable");
    }

    static AlchemyAssertion<Integer> validHttpStatusCode()
    {
        /*
         * See https://www.whoishostingthis.com/resources/http-status-codes
         */
        return greaterThanOrEqualTo(100).and(lessThanOrEqualTo(600));
    }

    /*
     * TODO: Add check to see if the class structure is that of a POJO.
     */
    static <Response> AlchemyAssertion<Class<Response>> validResponseClass()
    {
        return klass ->
        {
            checkThat(klass).isA(notNull());

            if (klass == Void.class)
            {
                throw new FailedAssertionException("Response class cannot be Void");
            }
        };
    }

    static AlchemyAssertion<HttpRequest> ready()
    {
        return request ->
        {
            checkThat(request)
                    .usingMessage("Request missing")
                    .isA(notNull());

            checkThat(request.method())
                    .usingMessage("Request missing HTTP Method")
                    .isA(notNull());

            checkThat(request.url())
                    .usingMessage("Request missing URL")
                    .isA(notNull());

            checkThat(request.url().getProtocol())
                    .isA(stringBeginningWith("http"));
        };
    }

    static AlchemyAssertion<String> validContentType()
    {
        return contentType ->
        {
            checkThat(contentType)
                    .usingMessage("missing Content-Type")
                    .isA(nonEmptyString());

            if (contentType.contains(ContentTypes.APPLICATION_JSON))
            {
                return;
            }

            if (contentType.contains(ContentTypes.PLAIN_TEXT))
            {
                return;
            }

            throw new FailedAssertionException("Not a valid JSON content Type: " + contentType);
        };
    }

    static AlchemyAssertion<HttpRequest> validRequest()
    {
        return request ->
        {
            checkThat(request)
                    .usingMessage("missing HTTP Request")
                    .isA(notNull());

            checkThat(request.url())
                    .usingMessage("missing request URL")
                    .isA(notNull());
        };
    }

    static AlchemyAssertion<JsonElement> jsonArray()
    {
        return json ->
        {
            checkThat(json).isA(notNull());

            if (!json.isJsonArray())
            {
                throw new FailedAssertionException("Expecting JSON Array, instead: " + json);
            }
        };
    }

    static AlchemyAssertion<HttpResponse> okResponse()
    {
        return response ->
        {
            checkThat(response).isA(notNull());

            if (!response.isOk())
            {
                throw new FailedAssertionException("Http Response not OK. Status Code: " + response.statusCode());
            }
        };
    }
}

File: src/main/java/tech/sirwellington/alchemy/http/HttpAssertions.kt
/*
 * Copyright © 2019. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.http

import com.google.gson.JsonElement
import tech.sirwellington.alchemy.annotations.access.Internal
import tech.sirwellington.alchemy.annotations.access.NonInstantiable
import tech.sirwellington.alchemy.arguments.AlchemyAssertion
import tech.sirwellington.alchemy.arguments.Arguments.checkThat
import tech.sirwellington.alchemy.arguments.FailedAssertionException
import tech.sirwellington.alchemy.arguments.assertions.*

/**
 * @author SirWellington
 */
@Internal
@NonInstantiable
internal object HttpAssertions
{

    @JvmStatic
    fun validHttpStatusCode(): AlchemyAssertion<Int>
    {
        /*
        * See https://www.whoishostingthis.com/resources/http-status-codes
        */
        return greaterThanOrEqualTo(100).and(lessThanOrEqualTo(600))
    }

    /*
    * TODO: Add check to see if the class structure is that of a POJO.
    */
    @JvmStatic
    fun <Response> validResponseClass(): AlchemyAssertion<Class<Response>>
    {

        return AlchemyAssertion { klass ->

            checkThat(klass).isA(nonNullReference())

            checkThat(klass == Void::class.java)
                    .usingMessage("Response class cannot be null")
                    .isA(falseStatement())
        }
    }

    @JvmStatic
    fun ready(): AlchemyAssertion<HttpRequest>
    {
        return AlchemyAssertion { request ->

            checkThat(request)
                    .usingMessage("Request missing")
                    .isA(nonNullReference())

            checkThat(request.method)
                    .usingMessage("Request missing HTTP Method")
                    .isA(nonNullReference())

            checkThat(request.url)
                    .usingMessage("Request missing URL")
                    .isA(nonNullReference())

            checkThat(request.url!!.protocol)
                    .isA(stringBeginningWith("http"))
        }
    }

    @JvmStatic
    fun validContentType(): AlchemyAssertion<String>
    {
        return AlchemyAssertion assertion@ { contentType ->

            checkThat(contentType)
                    .usingMessage("missing Content-Type")
                    .isA(nonEmptyString())

            if (contentType.contains(ContentTypes.APPLICATION_JSON))
            {
                return@assertion
            }

            if (contentType.contains(ContentTypes.PLAIN_TEXT))
            {
                return@assertion
            }

            throw FailedAssertionException("Not a valid JSON content Type: " + contentType)
        }
    }

    @JvmStatic
    fun validRequest(): AlchemyAssertion<HttpRequest>
    {
        return AlchemyAssertion { request ->

            checkThat(request)
                    .usingMessage("missing HTTP Request")
                    .isA(nonNullReference())

            checkThat(request.url)
                    .usingMessage("missing request URL")
                    .isA(nonNullReference())
        }
    }

    @JvmStatic
    fun jsonArray(): AlchemyAssertion<JsonElement>
    {
        return AlchemyAssertion { json ->

            checkThat<JsonElement>(json).isA(nonNullReference())

            if (!json.isJsonArray)
            {
                throw FailedAssertionException("Expecting JSON Array, instead: " + json)
            }
        }
    }

    @JvmStatic
    fun okResponse(): AlchemyAssertion<HttpResponse>
    {
        return AlchemyAssertion { response ->

            checkThat(response).isA(nonNullReference())

            if (!response.isOk)
            {
                throw FailedAssertionException("Http Response not OK. Status Code: " + response.statusCode())
            }
        }
    }

}
